qiskit_experiments/data_processing/data_processor.py

```diff
@@ -168,3 +186,75 @@ def train(self, data: List[Dict[str, Any]]):
                         train_data.append(self._call_internal(datum, call_up_to_node=index)[0])
 
                     node.train(train_data)
+
+    def _data_extraction(self, datum: Union[Dict, List[Dict]], call_up_to_node: int = None) -> Any:
+        """Extracts the data on which to run the nodes.
+
+        If the datum is a list of dicts then the data under self._input_key is extracted
+        from each dict and appended to a list which therefore contains all the data. If the
+        data processor has to_array set to True then the list will be converted to a numpy
+        array.
+
+        Args:
+            datum: Either a dict with the data under self._input_key or a list of such dicts.
+            call_up_to_node: The data processor will use each node in the processing chain
+                up to the node indexed by call_up_to_node. If this variable is not specified
+                then all nodes in the data processing chain are considered.
+
+        Returns:
+            The data formatted in such a way that it is ready to be processed by the nodes.
+
+        Raises:
+            DataProcessorError:
+                - If the input datum is not a list or a dict.
+                - If the data processor received a single datum but requires all the data to
+                  process it properly.
+                - If the input key of the data processor is not contained in the data.
+        """
+        data = None
+
+        if isinstance(datum, list):
+            for datum_ in datum:
+
+                if self._input_key not in datum_:
+                    raise DataProcessorError(
+                        f"The input key {self._input_key} was not found in the input datum."
+                    )
+
+                if data is None:
+                    data = [datum_[self._input_key]]
+                else:
+                    data.append(datum_[self._input_key])
+
+        elif isinstance(datum, dict):
+            if self.requires_all_data(call_up_to_node):
```

Comment:
I don't think this active error handling is necessary. We may want to feed an experimental result of single circuit into IQ processor node and want to normalized it. I prefer data processor to take always list of data dictionary and `len=1` can be processed as _all data_ rather than getting an error.

Reply:
See here: 057c38c